### PR TITLE
LibJS: Implement SuperCall for the Bytecode-VM

### PR DESCRIPTION
**Pull request by Hendiadyoin1**
So apperantly we dont do some other things right, which were previously hidden behind this like us messing up bindings in 
`automatic-semicolon-insertion.js:more break and continue asi` -> `InternalError: Lexical environment already has binding 'counter'`

### CODE REVIEW

File: Userland/Libraries/LibJS/Bytecode/ASTCodegen.cpp [hunk 1]
```diff
@@ -521,6 +521,22 @@ Bytecode::CodeGenerationErrorOr<void> Identifier::generate_bytecode(Bytecode::Ge
     return {};
 }
 
+Bytecode::CodeGenerationErrorOr<void> SuperCall::generate_bytecode(Bytecode::Generator& generator) const
+{
+    Vector<Bytecode::Register> argument_registers;
```
Comment (davidot) [inline Userland/Libraries/LibJS/Bytecode/ASTCodegen.cpp:526]: This is not correct for the (new) synthetic super call variant.
Which we use when you don't supply a constructor but have a parent class.
It should behave mostly like
```js
constructor(...args) {
    super(...args);
}
```

However we can't use that as you can poison `...args` by overwriting something in Array.

I don't think we need this right now so you could just check for it and return an error something like:
```cpp
return Bytecode::CodeGenerationError {
    this,
    "Synthetic super call not supported yet"sv
};
```

File: Userland/Libraries/LibJS/Bytecode/Op.cpp [hunk 2]
```diff
@@ -525,6 +526,50 @@ ThrowCompletionOr<void> Call::execute_impl(Bytecode::Interpreter& interpreter) c
...
+    // 3. Let func be GetSuperConstructor().
+    auto* func = get_super_constructor(vm);
+
+    // 4. Let argList be ? ArgumentListEvaluation of Arguments.
+    MarkedVector<Value> arg_list { vm.heap() };
+    for (size_t i = 0; i < m_argument_count; ++i)
+        arg_list.append(interpreter.reg(m_arguments[i]));
+
+    // 5. If IsConstructor(func) is false, throw a TypeError exception.
+    if (!func || !Value(func).is_constructor())
```
Comment (davidot) [inline Userland/Libraries/LibJS/Bytecode/Op.cpp:548]: You shouldn't need to check `!func` as `Value(nullptr)` should give a value of null and that isn't a constructor.
But honestly this is probably faster so keeping it is fine as well.

**Review by davidot — COMMENTED** [review]
LGTM but I'm not very well versed in bytecode.
I have just one comment about synthetic super calls.
And one tiny fix although this is probably faster so keeping it is fine.

File: Userland/Libraries/LibJS/Bytecode/Op.cpp (hunk 2 above)
Comment (Hendiadyoin1) [reply Userland/Libraries/LibJS/Bytecode/Op.cpp:548]: This is copied from the AST-interpreter

Comment (linusg) [reply Userland/Libraries/LibJS/Bytecode/Op.cpp:548]: This isn't the happy path anyway, so let's remove the `!func`

Comment (Hendiadyoin1) [reply Userland/Libraries/LibJS/Bytecode/Op.cpp:548]: Sure, will do

File: Userland/Libraries/LibJS/Bytecode/ASTCodegen.cpp (hunk 1 above)
Comment (Hendiadyoin1) [reply Userland/Libraries/LibJS/Bytecode/ASTCodegen.cpp:526]: I think I can manage to tinker something that gets this done with what is already written in the AST version

File: Userland/Libraries/LibJS/Bytecode/Op.cpp [hunk 3]
```diff
@@ -525,6 +526,58 @@ ThrowCompletionOr<void> Call::execute_impl(Bytecode::Interpreter& interpreter) c
...
+
+    // 5. If IsConstructor(func) is false, throw a TypeError exception.
+    if (!Value(func).is_constructor())
+        return vm.throw_completion<TypeError>(ErrorType::NotAConstructor, "Super constructor");
+
+    // 6. Let result be ? Construct(func, argList, newTarget).
+    auto* result = TRY(construct(vm, static_cast<FunctionObject&>(*func), move(arg_list), &new_target.as_function()));
+
+    // 7. Let thisER be GetThisEnvironment().
+    auto& this_er = verify_cast<FunctionEnvironment>(get_this_environment(vm));
```
Comment (linusg) [inline Userland/Libraries/LibJS/Bytecode/Op.cpp:563]: `this_environment`, let's not use the awful spec name here

File: Userland/Libraries/LibJS/Bytecode/Op.cpp [hunk 4]
```diff
@@ -525,6 +526,58 @@ ThrowCompletionOr<void> Call::execute_impl(Bytecode::Interpreter& interpreter) c
...
+
+    // 8. Perform ? thisER.BindThisValue(result).
+    TRY(this_er.bind_this_value(vm, result));
+
+    // 9. Let F be thisER.[[FunctionObject]].
+    // 10. Assert: F is an ECMAScript function object.
+    // NOTE: This is implied by the strong C++ type.
+    auto& f = this_er.function_object();
```
Comment (linusg) [inline Userland/Libraries/LibJS/Bytecode/Op.cpp:571]: This should be after 9, 10 + the note can be standalone

File: Userland/Libraries/LibJS/Bytecode/Op.cpp [hunk 5]
```diff
@@ -1008,6 +1061,24 @@ String Call::to_string_impl(Bytecode::Executable const&) const
...
+    if (m_is_synthetic) {
+        builder.append(" arguments:[...acc]"sv);
+    } else if (m_argument_count != 0) {
+        builder.append(" arguments:["sv);
+        for (size_t i = 0; i < m_argument_count; ++i) {
+            builder.appendff("{}", m_arguments[i]);
+            if (i != m_argument_count - 1)
+                builder.append(',');
+        }
+        builder.append(']');
```
Comment (linusg) [inline Userland/Libraries/LibJS/Bytecode/Op.cpp:1077]: String::join pls

**Review by linusg — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibJS/Bytecode/Op.cpp (hunk 4 above)
Comment (Hendiadyoin1) [reply Userland/Libraries/LibJS/Bytecode/Op.cpp:571]: This is from the AST version, but I can move the Note

File: Userland/Libraries/LibJS/Bytecode/Op.cpp (hunk 5 above)
Comment (Hendiadyoin1) [reply Userland/Libraries/LibJS/Bytecode/Op.cpp]: Turns out the compiler error is just not helpful
(It complains about the `incomplete type 'const JS::Bytecode::Register[]'` and not about the missing const you would have gotten after the type decay)

File: Userland/Libraries/LibJS/Bytecode/Op.cpp [hunk 6]
```diff
@@ -525,6 +526,59 @@ ThrowCompletionOr<void> Call::execute_impl(Bytecode::Interpreter& interpreter) c
...
+    VERIFY(new_target.is_object());
+
+    // 3. Let func be GetSuperConstructor().
+    auto* func = get_super_constructor(vm);
+
+    // 4. Let argList be ? ArgumentListEvaluation of Arguments.
+    MarkedVector<Value> arg_list { vm.heap() };
+    if (m_is_synthetic) {
+        auto const& value = interpreter.accumulator();
+        auto const& array_value = static_cast<Array const&>(value.as_object());
```
Comment (davidot) [inline Userland/Libraries/LibJS/Bytecode/Op.cpp:546]: Could add a VERIFY to make sure `value` is actually an Array.
```cpp
VERIFY(value.is_object() && is<Array>(value.as_object()));
```

**Pull request merged**